feat: Add measurement emplace functions

stephenswat opened the pull request:
Recently, the measurement EDM was replaced by a memory-optimal bulk storage, but this makes it a bit harder to add new measurements to a collection. This commit adds two new emplace functions which variadically allow users to add measurements, similar to how they might use `std::vector::emplace_back`.

andiwand commented on Examples/Framework/include/ActsExamples/EventData/Measurement.hpp:
```diff
@@ -282,19 +289,56 @@ class MeasurementProxyBase {
+  ///
+  template <typename OtherDerived>
+  void construct(const OtherDerived& other)
+    requires(!ReadOnly) && requires {
+      { other.size() } -> std::integral;
+      { other.sourceLink() };
+      { other.subspaceIndexVector() };
+      { other.parameters() };
+      { other.covariance() };
+    }
```
1. should we create a named concept for this?
2. also what kind of object can be fed into this?

stephenswat replied:
Basically just any measurement proxy.

andiwand commented on Examples/Framework/include/ActsExamples/EventData/Measurement.hpp:
```diff
@@ -485,4 +529,25 @@ class VariableMeasurementProxy
   }
 };
 
+template <typename... Args>
+MeasurementContainer::VariableProxy
+MeasurementContainer::emplaceVariableMeasurement(std::uint8_t size,
+                                                 Args&&... args) {
+  VariableProxy meas = makeMeasurement(size);
+
+  meas.construct(std::forward<Args>(args)...);
+
+  return meas;
+}
+
+template <std::size_t Size, typename... Args>
+MeasurementContainer::FixedProxy<Size>
+MeasurementContainer::emplaceFixedMeasurement(Args&&... args) {
+  FixedProxy<Size> meas = makeMeasurement<Size>();
+
+  meas.construct(std::forward<Args>(args)...);
+
+  return meas;
+}
```
I wonder if it might be easier to inline define these functions

stephenswat replied:
I don't think that's possible, because the actual types for `FixedProxy` and `VariableProxy` are defined only _after_ the class definition, so you run into instantiations of undefined templates. :(

andiwand reviewed (commented):
Looks nice 👍 Two more requests
1. Could you add tests for this interface?
2. Would you be up to change the existing measurements creation in the Examples to the new interface?

stephenswat commented:
I went and:

1. Added tests for these new methods;
2. Converted all instances of `makeMeasurement` to these new methods;
3. Added a named measurement concept.

andiwand commented on Examples/Framework/include/ActsExamples/EventData/Measurement.hpp:
```diff
@@ -117,6 +119,12 @@ class MeasurementContainer {
     return getMeasurement<Size>(addMeasurement(Size));
   }
 
+  template <typename... Args>
+  VariableProxy emplaceVariableMeasurement(std::uint8_t size, Args&&... args);
+
+  template <std::size_t Size, typename... Args>
+  FixedProxy<Size> emplaceFixedMeasurement(Args&&... args);
```
If we use concrete parameter types here we could overload an `emplaceFixedMeasurement` function for both cases. Nothing critical but that might align more with the rest of the API

andiwand replied:
What I mean is that there could be a single function called `emplace` or `emplaceMeasurement` and one has the overload with the size template for fixed size proxy returns. But I think that only works if you type out the parameters otherwise C++ will not let us overload it.

stephenswat replied:
Ah right! Let me see if I can get that to work.

stephenswat replied:
Wow, that went surprisingly easily.

andiwand commented on Examples/Framework/include/ActsExamples/EventData/Measurement.hpp:
```diff
@@ -282,19 +290,49 @@ class MeasurementProxyBase {
+  /// @brief Construct the measurement from a sourcelink, subspace vector,
+  /// parameters, and covariance.
+  ///
+  template <typename Subspace, typename ParameterDerived,
+            typename CovarianceDerived>
+  void construct(const Acts::SourceLink& source_link, Subspace&& subspace,
+                 const Eigen::DenseBase<ParameterDerived>& parameters,
+                 const Eigen::DenseBase<CovarianceDerived>& covariance)
+    requires(!ReadOnly) && requires {
+      { this->self().setSubspaceIndices(std::forward<Subspace>(subspace)) };
+      { this->self().parameters() = parameters };
+      { this->self().covariance() = covariance };
+    }
+  {
+    setSourceLink(source_link);
+    self().setSubspaceIndices(std::forward<Subspace>(subspace));
+    self().parameters() = parameters;
+    self().covariance() = covariance;
+  }
```
I wonder if `construct` is a good name for the function as it does not really construct something but rather fill all the information? Maybe `fill` makes more sense? Again not a real issue

stephenswat replied:
Good idea, I like "fill".

stephenswat replied:
Done.

andiwand reviewed (approved):
Nice! Two small remarks but nothing crucial

andiwand reviewed (approved):
nice! thank you!

acts-project-service commented:
## :red_circle: Athena integration test results [a129eebc1c617d7988da249f3929ff0dfb4394d5]
### Build job with this PR failed!
Please investigate the [build job](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/43984435) for the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/8207882)!
